### ISSUE TITLE
Crossing coordinate atlas parcellations, resting-state networks, and specific brain regions?

### ISSUE THREAD

**Issue by jlhanson5** [dPys/PyNets#143]
* PyNets version:0.7.25
* Python version: 3.6.6 (Anaconda)
* Operating System: Ubuntu 16.04.4

I had some (hopefully) quick, related-questions

1) First, using the '-n' flag, one can pull Yeo-Schaefer networks, but some experimenting suggest that certain atlases (called via the '-a' flag) have different coverages of these networks. Is there a good way to think about how these two options intersect? Some -a specification seem to yield "better coverage" of some networks (just experimenting with the default mode, etc.)?

2) Related, if one were to use another atlas (that don't look to be preconfigured in regards to the '-a' flag) such as AICHAJoliot2015 or BrainnetomeAtlasFan2016 included in the pynets/atlas folder, how can you pull out specific networks (or will PyNets automatically work through the appropriate coordinates)? 

3) And finally, if one wanted to focus on a small set of regions (say, for example, pull out the centrality of an ACC ROI from an atlas), could that be accomplished in PyNet? Or would it make more sense to generate graph files and then port them to igraph or other programs, etc?

Pardon the slightly basic questions. Based on the existing documentation, I wasn't sure about things. I know there's a new user-guide coming, I was just very excited about getting PyNets implemented (and getting more data analyzed!). 

Thanks much in advance. 
Jamie.

**Comment by dPys** [comment]
Hi @jlhanson5 ,

So sorry for the delayed reply! Just wrapped up a Nipype sprint (trying to find ways to optimize the PyNets workflow among other things)

I'll do my best to answer your questions:
1. Your observation is correct and underscores some of the core concerns that PyNets hopes to address. In short, there _is_ considerable intersection between coverage of the Yeo-Schaefer networks (RSN's) and the anatomically-defined atlases available through the -a flag. The way the RSN's are isolated is actually pretty simple with the current implementation-- each node of a given atlas is classified as either falling within or outside of the spatial boundaries of the RSN (as defined by the Yeo-Schaeffer templates) in question; that is, the problem is currently being treated as boolean. These boundaries are also extended to also include a "neighborhood" of possible node members. In the case where spherical nodes are specified, a default of 4 x 4 x 4 voxel euclidian margin of error is used to define that neighborhood. Likewise, a >75% spatial overlap is required for a parcel to be classified as belonging to the neighborhood (i.e. when using the -parc flag). In the near future, I will be making this value a user-configurable parameter... But do know that this approach to defining RSN's is very likely to evolve into routines that are much more intricate. See recent [Tom Yeo's](https://academic.oup.com/cercor/advance-article-abstract/doi/10.1093/cercor/bhy123/5033556) take on this if you haven't already.
That all being said, as a rule of thumb-- if an atlas has fewer, smaller parcels overall, then fewer parcels will probably be classified as "belonging" to the RSN. The converse is also the case.

2. PyNets will automatically work through the appropriate coordinates :)

3. You should be able to get whatever local/nodal metrics you want from the pynets outputs (FYI: the *_neat.csv file is the one that organizes the results into a clean database). There are multiple forms of centrality measures that are produced for each atlas node. To figure out which measure belongs to which node, they should be numerically indexed according to the labeling scheme (e.g. ACC would be nodes 35 & 36 if using the whole-brain [AAL atlas](https://github.com/dPys/PyNets/blob/master/pynets/labelcharts/AAL2zourioMazoyer2002.txt) dictionary)

**Comment by jlhanson5** [comment]
Ah thanks for this help! I tried the following syntax--

`pynets_run.py -i /home/jamielh/Volumes/Hanson/NKI_HealthyBrainNetwork/SI/R1/derivatives/fmriprep/sub-NDARAE199TDD/func/sub-NDARAE199TDD_task-resting_bold_space-MNI152NLin2009cAsym_variant-smoothAROMAnonaggr_preproc.nii.gz -id NDARAE199TDD_temp -ua /home/jamielh/Volumes/Hanson/Neuroimaging_Resources/pynets_data/atlases/AICHAJoliot2015.nii.gz,/home/jamielh/Volumes/Hanson/Neuroimaging_Resources/pynets_data/atlases/BrainnetomeAtlasFan2016.nii.gz -n Limbic -ns 2 -mod partcorr,sps -plt`

but ran into some challenges. I start to get nan for different graph paramaters and many others are just 1s. 

I uploaded 4 different related crash logs here--
[ONE](https://www.dropbox.com/s/m9uk3rj68qwqlyr/crash-20181012-160748-jamielh-_plot_all_node0-3122a1b8-5233-4085-9f7c-d0e91ac32256.txt?dl=0
)
[TWO](https://www.dropbox.com/s/nawq0fp85tkrwm1/crash-20181012-160748-jamielh-_plot_all_node1-13dab831-c10a-4463-9fc9-2f0d8cc9c2e0.txt?dl=0
)
[THREE](https://www.dropbox.com/s/6j78hucdvx9tuju/crash-20181012-160748-jamielh-_plot_all_node2-92280eca-2084-44f4-978c-ffb8cc6946c5.txt?dl=0
)
[FOUR](https://www.dropbox.com/s/37kdjlnstm8m0c9/crash-20181012-160748-jamielh-_plot_all_node3-070f5d7a-961d-483d-aefc-617eccd49028.txt?dl=0
)

In brief, I tried to use the AICHAJoliot2015 and BrainnetomeAtlasFan2016 to calculate (Yeo) Limbic network properties. I don't think I screwed up the syntax (as I could get some other atlases to complete)... but I'm likely doing something else in error. 

Thoughts? Thanks much! (And I'll be in touch off the repo, but, ideally, I would invite you to be a co-author if anything emerged from my pyneting... if you are advisers are cool with it, etc. Developing tools is too much work to not get at least _some_ pubs out of it!)

Cheers,
Jamie.

**Comment by dPys** [comment]
Hi @jlhanson5 , 

Thanks for reporting. Your command is correct :)

Two follow-up questions-- 
1) What happens if you drop the -plt flag? does it still error?
2) What happens if you instead use the Power atlas. Does it still error then?

An empty distance matrix either means that these particular atlases didn't yield any nodes when using the Limbic restricted network, or that a length matrix could not be inferred from the basic adj. matrix for a mathematical reason.

-Derek

**Comment by jlhanson5** [comment]
> What happens if you drop the -plt flag? does it still error?
It finishes up fine, actually

> What happens if you instead use the Power atlas. Does it still error then?
 Using -plt, it crashes; without -plt, it seems fine.

**Comment by dPys** [comment]
Thanks for reporting! Very odd that this is happening with -plt, but not without. Perhaps try a different functional image? I am unable to reproduce the error


![997_aichajoliot2015_sps_limbic_0 2_4mm_nosm_func_glass_viz](https://user-images.githubusercontent.com/16432683/46912394-15776880-cf3a-11e8-863e-68276e52bfbc.png)


Also try pip install pynets==0.7
![997_brainnetomeatlasfan2016_sps_limbic_0 2_4mm_nosm_func_glass_viz](https://user-images.githubusercontent.com/16432683/46912395-1c9e7680-cf3a-11e8-93f8-aef65f240630.png)
.26

**Comment by jlhanson5** [comment]
I upgraded to 0.7.26 and tried it on 3 subs, but no luck. Other thoughts? Should I try to upgrade pandas (or the other packages used with -plt)? I think I have pandas 0.21.1 currently installed.

**Comment by jlhanson5** [comment]
Perfect, I uploaded 2 subjects here--

[subject1](https://www.dropbox.com/s/ctm8mc5tk4bm6gm/sub-NDARZG263HRK_task-resting_bold_space-MNI152NLin2009cAsym_variant-smoothAROMAnonaggr_preproc.nii.gz?dl=0
)
[subject2](https://www.dropbox.com/s/jpkmatkpx7iu82z/sub-NDARZT772PU4_task-resting_bold_space-MNI152NLin2009cAsym_variant-smoothAROMAnonaggr_preproc.nii.gz?dl=0
)

Both were just run through fmriprep (using most of their standard non-Freesurfer settings and with ICA-AROMA). Tell me your thoughts.

**Comment by dPys** [comment]
Hi @jlhanson5 , so I took a look, and a couple things:

1) You're going to want to apply a brain mask to your data. bet would be the easiest way to do this (with the -m flag), followed by fslmaths with the -mas option.
2) There should be a confound regressor file that gets output from fmriprep. You are going to want to include this is your pynets run using the -conf flag.
3) If the above steps do not resolve the issue, then I recommend imposing a higher density on the graph using something like: -dt -thr 0.3 

Basically the issue is that either due to noise in the time-series (correctable by including a confound regressor file) or due to a lack of connectivity in the limbic circuit for that image, or both, the resulting connectivity matrix is excessively sparse and hence cannot be plotted.

derek;

**Comment by dPys** [comment]
Closing this issue, but feel free to post again to reopen at any time.

Cheers,
Derek
